service/history/shard/context.go

```diff
@@ -176,16 +176,21 @@ var _ Context = (*contextImpl)(nil)
 var (
 	// ErrShardClosed is returned when shard is closed and a req cannot be processed
 	ErrShardClosed = errors.New("shard closed")
+
+	// ErrShardRecentlyClosed is returned when a shard has recently been closed, this
+	// error is transient, and should not cause error logs and error metrics to be emitted
+	ErrShardRecentlyClosed = errors.New("shard recently closed")
```

Comment:
There are other places (history queues) that stop themselves when they see shard closed error. This change will delay that process and might cause more attempts on recently closed shards. 
See usages of `shard.ErrShardClosed`

Reply:
Yeah, I see, I should have found that. 

Maybe it's better to add a timestamp to the shardClosed error, then the places that care about it can act accordingly?

Reply:
So I'd suggest creating a shardClosed error type, this would have the same message, but include the timestamp of when the shard was closed. 
Then I will change the places that use the error to check the type with errors.As/errors.Is, and then I can extract the timestamp in the handler where I need it. 

Do you think that would work? (just want to confirm before I start working on it :slightly_smiling_face:  )

Reply:
Not a bad idea to include shard closed timestamp for informational purposes.
However I don't see a reason to check the time elapsed since shard closed. Just change the task.go with
```
if err == shard.ErrShardClosed {
	return err
}
```

We should not keep retry on it. Doesn't matter when it's closed. it will not re-open :) 
cc @Shaddoll

Reply:
So I added the context in the alert. 

We have seen the bad state persist for hours, so I'm worried about just removing the metrics completely. 

I'm not sure what causes it, but we just keep seeing these errors emitted for some shard/host combination, until the host is restarted. 

So we should handle that as a "real" error state and emit the logs, etc. 

If we can resolve this, then we should probably not need this timeout any more.